### PR TITLE
Olelensmar/feat/helm chart form editors

### PR DESCRIPTION
**Pull request by olensmar**
This PR adds form editors and source-editor validation for Helm Chart.yaml files and values files that have a corresponding values.schema.json file

Helm Chart schema and UI-schema descriptions were taken from https://helm.sh/docs/topics/charts/#the-chartyaml-file

## Changes

-

## Fixes

-

## How to test it

-

## Screenshots

image

image

image

## Checklist

- [X] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: src/redux/services/fileEntry.ts [hunk 1]
```diff
@@ -195,6 +195,14 @@ export function getAbsoluteResourcePath(resource: K8sResource, fileMap: FileMapT
   return path.join(fileMap[ROOT_FILE_ENTRY].filePath, resource.filePath);
 }
 
+/**
+ * Returns the absolute path to the file that containing specified resource
```
Comment (devcatalin) [inline src/redux/services/fileEntry.ts:199]: this comment should be modified

File: src/redux/services/schema.ts [hunk 2]
```diff
@@ -58,6 +68,24 @@ export function getResourceSchema(resource: K8sResource) {
   return schemaCache.get(resource.kind);
 }
 
+export function getSchemaForPath(filePath: string, fileMap: FileMapType): any | undefined {
+  if (filePath?.toLowerCase().endsWith('/chart.yaml')) {
```
Comment (devcatalin) [inline src/redux/services/schema.ts:72]: should use `path.sep` here

File: src/hooks/useResourceYamlSchema.ts [hunk 3]
```diff
@@ -30,7 +34,7 @@ function useResourceYamlSchema(yaml: any, resource: K8sResource | undefined) {
         enableSchemaRequest: true,
         hover: true,
         completion: true,
-        isKubernetes: true,
+        isKubernetes: resource,
```
Comment (devcatalin) [inline src/hooks/useResourceYamlSchema.ts:37]: should this be `Boolean(resource)` ? but it's maybe redundant

Comment (devcatalin) [reply src/hooks/useResourceYamlSchema.ts:37]: I just saw that `| undefined` was removed from the `resource` argument, so wouldn't this be always true?

File: src/components/organisms/ActionsPane/ActionsPane.tsx [hunk 4]
```diff
@@ -341,7 +342,12 @@ const ActionsPane = (props: {contentHeight: string}) => {
   }, [selectedResourceId, resourceMap]);
 
   useEffect(() => {
```
Comment (devcatalin) [inline src/components/organisms/ActionsPane/ActionsPane.tsx:344]: missing `selectedPath` as dependency in this `useEffect`

File: src/redux/services/schema.ts (hunk 2 above)
Comment (devcatalin) [reply src/redux/services/schema.ts:72]: or use `path.basename(filePath).toLowerCase() === 'chart.yaml'`

**Comment by olensmar** [comment]
now with a ui-schema for the helm-chart: 
image

File: src/components/organisms/ActionsPane/ActionsPane.tsx (hunk 4 above)
Comment (olensmar) [reply src/components/organisms/ActionsPane/ActionsPane.tsx:344]: adding this breaks the source editor (it doesn't show up - not sure why)

File: resources/form-schemas/helm-chart-schema.json [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    },
+    "maintainers": {
+      "type": "array",
+      "items":
+      {
+        "type": "object",
+        "properties": {
+          "name": {
+            "type": "string",
+            "description": "The maintainers name"
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:108]: small type here and below, should be `maintainer's`

File: resources/form-schemas/helm-chart-schema.json [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+          "import-values": {
+            "type": "array",
+            "description": "ImportValues holds the mapping of source values to parent key to be imported. Each item can be a string or pair of child/parent sublist items.",
+            "items":
+            {
+              "type": "string"
+            }
+          },
+          "alias": {
+            "description": "Alias to be used for the chart. Useful when you have to add the same chart multiple times",
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:90]: what does "when you have to add the same chart multiple times" mean?

File: src/components/organisms/ActionsPane/ActionsPane.tsx (hunk 4 above)
Comment (olensmar) [reply src/components/organisms/ActionsPane/ActionsPane.tsx:344]: sorry - my bad - you're right!

File: resources/form-schemas/helm-chart-schema.json [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            "type": "array",
+            "description": "Tags can be used to group charts for enabling/disabling together",
+            "items":
+            {
+              "type": "string"
+            }
+          },
+          "import-values": {
+            "type": "array",
+            "description": "ImportValues holds the mapping of source values to parent key to be imported. Each item can be a string or pair of child/parent sublist items.",
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:83]: maybe this is easier to read: `ImportValues defines the source value to parent key mappings that will be imported.`

File: resources/form-schemas/helm-chart-schema.json [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            "type": "string",
+            "description": "The version of the chart (\"1.2.3\")"
+          },
+          "repository": {
+            "type": "string",
+            "description": "The repository URL (\"https://example.com/charts\") or alias (\"@repo-name\")"
+          },
+          "condition": {
+            "type": "string",
+            "description": "A yaml path that resolves to a boolean, used for enabling/disabling charts (e.g. subchart1.enabled )"
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:71]: I suggest: `yaml object path`

File: resources/form-schemas/helm-chart-schema.json [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+      {
+        "type": "object",
+        "properties": {
+          "name": {
+            "type": "string",
+            "description": "The name of the chart (nginx)"
+          },
+          "version": {
+            "type": "string",
+            "description": "The version of the chart (\"1.2.3\")"
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:63]: let's add `e.g.` at the start of the examples for all properties

File: resources/form-schemas/helm-chart-schema.json [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+      }
+
+    },
+    "home": {
+      "type": "string",
+      "description": "The URL of this projects home page"
+    },
+    "sources": {
+      "type": "array",
+      "description": "A list of URLs to source code for this project",
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:44]: I suggest: `list of URLs pointing to the source code of this...`

File: resources/form-schemas/helm-chart-schema.json [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+      "description": "A list of keywords about this project",
+      "items":
+      {
+        "type": "string"
+      }
+
+    },
+    "home": {
+      "type": "string",
+      "description": "The URL of this projects home page"
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:40]: small typo: `project's`

File: resources/form-schemas/helm-chart-schema.json [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+      "type": "string",
+      "description": "A URL to an SVG or PNG image to be used as an icon"
+    },
+    "appVersion": {
+      "type": "string",
+      "description": "The version of the app that this contains. Needn't be SemVer. Quotes recommended."
+    },
+    "deprecated": {
+      "type": "boolean",
+      "description": "Whether this chart is deprecated"
```
Comment (devcatalin) [inline resources/form-schemas/helm-chart-schema.json:134]: add "or not" at the end?

File: src/components/molecules/FormEditor/FormEditor.tsx [hunk 13]
```diff
@@ -122,25 +127,44 @@ const FormEditor = (props: {formSchema: any; formUiSchema?: any}) => {
...
       if (selectedResource) {
-        let formString = stringify(formData);
         const content = mergeManifests(selectedResource.text, formString);
-
         if (content.trim() !== selectedResource.text.trim()) {
           dispatch(updateResource({resourceId: selectedResource.id, content}));
         }
+      } else if (selectedPath) {
+        try {
+          const filePath = getAbsoluteFilePath(selectedPath, fileMap);
```
Comment (devcatalin) [inline src/components/molecules/FormEditor/FormEditor.tsx:139]: I think for future consideration we should have selectors for these types of functions so we don't have to import the entire `fileMap` state into each components that needs to call a simple function like this one.
By using custom selectors, only the output of the selector will affect the rendering of the component, while in this case, any fileMap change will trigger a re-render of this component

**Review by devcatalin — CHANGES_REQUESTED** [review]

**Review by devcatalin — APPROVED** [review]
Looks really nice!

**Pull request merged**